js/components/MilkdropWindow/index.js

```diff
@@ -39,16 +40,15 @@ class MilkdropWindow extends React.Component {
       ["butterchurn-presets"],
       require => {
         const butterchurnPresets = require("butterchurn-presets");
-        const presets = butterchurnPresets.getPresets();
-        const presetKeys = Object.keys(presets);
+        this.presets = butterchurnPresets.getPresets();
+        this.presetKeys = Object.keys(this.presets);
+        this.presetHistory = [];
         this.cycleInterval = setInterval(() => {
-          const presetIdx = Math.floor(presetKeys.length * Math.random());
-          const preset = presets[presetKeys[presetIdx]];
-          // The visualizer may not have initialized yet.
-          if (this.visualizer != null) {
-            this.visualizer.loadPreset(preset, PRESET_TRANSITION_SECONDS);
-          }
+          this._nextPreset(PRESET_TRANSITION_SECONDS);
         }, MILLISECONDS_BETWEEN_PRESET_TRANSITIONS);
+        document.addEventListener("keydown", e => {
+          this._handleKeyboardInput(e);
+        });
```

Comment:
The reason we have to bind to the document, rather than being able to directly bind to the canvas, is that we are not actually setting the "focus" when a user selects a window. Does real Winamp only recognize these keys when Milkdrop is the selected window?

Reply:
ahh, makes sense.  Winamp requires milkdrop to have focus, I'm personally fine with them working always, and they don't seem to conflict, so I think it would be OK to leave it on document then.

Reply:
Oh, one more thing I just realized. We should remove this event listener inside `componentWillUnMount`. Maybe do 

```
this._handleKeyboardInput = this._handleKeyboardInput.bind(this);
```

inside the constructor and then you can just do:

```
document.addEventListener("keydown", this._handleKeyboardInput);
```

and

```
document.removeEventListener("keydown", this._handleKeyboardInput);
```

Reply:
oh yea, definitely, good catch